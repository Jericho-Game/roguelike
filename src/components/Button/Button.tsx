import { ComponentPropsWithoutRef, ElementType, PropsWithChildren } from 'react';
import classnames from 'classnames';

type ButtonProps<T extends ElementType> = PropsWithChildren<{
  as?: T;
  className?: string;
  variant: 'primary' | 'secondary' | 'alternate' | 'icon';
}> & Omit<ComponentPropsWithoutRef<T>, 'as' | 'className' | 'variant' | 'children'>;

export default function Button<T extends ElementType = 'button'>({
  as,
  variant,
  className = '',
  ...props
}: ButtonProps<T>) {
  const Component = as ?? 'button';
  return (
    <Component
      className={classnames(
        `btn btn-${variant}`,
        className,
        {
          [`hover:bg-${variant}-hover`]: variant !== 'alternate' && variant !== 'icon',
          'hover:text-gray-800 hover:border-gray-800': variant === 'alternate',
          'hover:text-gray-500 hover:bg-gray-100': variant === 'icon',
          'focus:outline-none focus:ring-2 focus:ring-inset focus:ring-secondary-hover': variant === 'icon',
        },
      )}
      {...props}
    />
  );
}
