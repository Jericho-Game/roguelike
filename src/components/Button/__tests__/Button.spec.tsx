import { ElementType } from 'react';
import { render } from '@testing-library/react';
import Button, { ButtonProps } from '../Button';

const defaultProps = {
  variant: 'primary' as const,
};

const create = (props = {}) => <Button {...defaultProps} {...props} />;

describe('Button Test', () => {
  it('matches snapshot', () => {
    const { container } = render(create());

    expect(container.firstChild).toMatchSnapshot();
  });

  it('as another button', () => {
    const as = ({ children, ...props }: ButtonProps<ElementType>) => (
      <div {...props}>{children}</div>
    );
    const { container } = render(create({ as }));

    expect(container.firstChild).toMatchSnapshot();
  });

  it('with className', () => {
    const className = 'test';
    const { container } = render(create({ className }));

    expect(container.firstChild).toHaveClass(className);
  });

  it('as icon variant', () => {
    const Icon: React.FC = (props) => <div {...props}>Icon</div>;
    const { container } = render(create({
      variant: 'icon',
      children: <Icon />,
    }));
    const button = container.firstChild;
    const icon = button?.firstChild;

    expect(button).toHaveClass('btn-icon');
    expect(icon).toHaveTextContent('Icon');
  });
});
