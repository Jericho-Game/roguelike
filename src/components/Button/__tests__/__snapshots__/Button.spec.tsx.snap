// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Test as another button 1`] = `
<div
  class="btn btn-primary hover:bg-primary-hover"
/>
`;

exports[`Button Test matches snapshot 1`] = `
<button
  class="btn btn-primary hover:bg-primary-hover"
/>
`;
